src/libgrate/matrix.c

```diff
@@ -143,7 +143,7 @@ void mat4_perspective(struct mat4 *m, float fov, float aspect,
 	m->xx = cotangent / aspect;
 	m->yy = cotangent;
 	m->zz = -(far + near) / depth;
-	m->zw = -1.0f;
-	m->wz = -2.0f * near * far / depth;
+	m->zw = -2.0f * near * far / depth;
+	m->wz = -1.0f;
```

Comment:
And what we're doing doen't seem to match column-major even if we interpret column-major as meaning memory layout (which seems to be what the OpenGL specification and [Wikipedia claims](https://en.wikipedia.org/wiki/Row-_and_column-major_order), so I guess that part of argument was flawed), because OpenGL 2.1 defines:

> LoadMatrix takes a pointer to a 4 × 4 matrix stored in column-major order as 16
> consecutive floating-point values, i.e. as
> a1 a5 a9 a13
> a2 a6 a10 a14
> a3 a7 a11 a15
> a4 a8 a12 a16

And:

> The arguments to
> void Translate{fd}( T x, T y, T z );
> give the coordinates of a translation vector as (x y z)^T . The resulting matrix is a translation by the specified vector:
> 1 0 0 x
> 0 1 0 y
> 0 0 1 z
> 0 0 0 1

Which shows that the matrix is intended to be the transpose of what i.e mat4_translate produces.

Also, for this particular case, the glFrustum (which is the closest thing defined in OpenGL) is defined as:

> The corresponding matrix is
> (2n) / (r−l), 0, (r+l) / (r−l), 0
> 0, (2n) / (t−b), (t+b) / (t−b), 0
> 0 0 -(f+n) / (f−n), -(2fn) / (f−n)
> 0, 0, −1, 0

Which, according to the order defined in LoadMatrix, would mean "wz" would be the right memory location for `-2.0f * near * far / depth`.¨